File: Zeze/Arch/ProviderDistributeVersion.cs
using DotNext.Collections.Generic;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Zeze.Services.ServiceManager;

namespace Zeze.Arch
{
    public class ProviderDistributeVersion
    {
        public Application Zeze { get; set; }
        public LoadConfig LoadConfig { get; set; }
        public Zeze.Net.Service ProviderService { get; set; }

        public Dictionary<long, ProviderDistribute> DistributeVersions { get; } = new();

        public void AddServer(ServiceInfo info)
        {
            DistributeVersions.GetOrAdd(info.Version, _ => new ProviderDistribute(info.Version, Zeze, ProviderService, LoadConfig)).AddServer(info);
        }

        public void RemoveServer(ServiceInfo info)
        {
            if (DistributeVersions.TryGetValue(info.Version, out var versions))
            {
                versions.RemoveServer(info);
                // 有线程问题，干脆不删除，保留空的在里面吧。
                //if (DistributeVersions.Count == 0)
                //    DistributeVersions.Remove(info.Version);
            }
        }

        public ProviderDistribute SelectDistribute(long version)
        {
            // 更多选择算法
            if (DistributeVersions.TryGetValue(version, out var versions))
                return versions;
            return null;
        }
    }
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Arch/ProviderDistributeVersion.java
package Zeze.Arch;

import Zeze.Application;
import Zeze.Net.Service;
import Zeze.Services.ServiceManager.BServiceInfo;
import Zeze.Util.FewModifyMap;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ProviderDistributeVersion {
	public final Application zeze;
	public final LoadConfig loadConfig;
	private final Service providerService;
	private final FewModifyMap<Long, ProviderDistribute> providerDistributes = new FewModifyMap<>();

	public ProviderDistributeVersion(Application zeze, LoadConfig loadConfig, Service providerService) {
		this.zeze = zeze;
		this.loadConfig = loadConfig;
		this.providerService = providerService;
	}

	public void addServer(@NotNull BServiceInfo info) {
		providerDistributes.computeIfAbsent(info.getVersion(),
				k -> new ProviderDistribute(zeze, loadConfig, providerService, k)).addServer(info);
	}

	public void removeServer(@NotNull BServiceInfo info) {
		var providerDistribute = providerDistributes.get(info.getVersion());
		if (providerDistribute != null)
			providerDistribute.removeServer(info);
	}

	public @Nullable ProviderDistribute selectDistribute(long version) {
		return providerDistributes.get(version);
	}
}
